src/aotcompile.cpp

```diff
@@ -1981,34 +1971,24 @@ void jl_get_llvmf_defn_impl(jl_llvmf_dump_t* dump, jl_method_instance_t *mi, siz
     jl_code_info_t *src = NULL;
     jl_code_instance_t *codeinst = NULL;
     JL_GC_PUSH3(&src, &jlrettype, &codeinst);
-    if (jl_is_method(mi->def.method) && mi->def.method->source != NULL && mi->def.method->source != jl_nothing && jl_ir_flag_inferred(mi->def.method->source)) {
-        // uninferred opaque closure
-        src = (jl_code_info_t*)mi->def.method->source;
-        if (src && !jl_is_code_info(src))
-            src = jl_uncompress_ir(mi->def.method, NULL, (jl_value_t*)src);
+    jl_value_t *ci = params.lookup(mi, world, world);
+    if (ci && ci != jl_nothing) {
+        codeinst = (jl_code_instance_t*)ci;
+        src = (jl_code_info_t*)jl_atomic_load_relaxed(&codeinst->inferred);
+        jlrettype = codeinst->rettype;
     }
-    else {
-        jl_value_t *ci = params.lookup(mi, world, world);
-        if (ci != jl_nothing) {
-            codeinst = (jl_code_instance_t*)ci;
+    if (!src || (jl_value_t*)src == jl_nothing) {
+        codeinst = jl_type_infer(mi, world, 0, SOURCE_MODE_FORCE_SOURCE);
+        if (codeinst) {
             src = (jl_code_info_t*)jl_atomic_load_relaxed(&codeinst->inferred);
-            if ((jl_value_t*)src != jl_nothing && !jl_is_code_info(src) && jl_is_method(mi->def.method))
-                src = jl_uncompress_ir(mi->def.method, codeinst, (jl_value_t*)src);
             jlrettype = codeinst->rettype;
-            codeinst = NULL; // not needed outside of this branch
-        }
-        if (!src || (jl_value_t*)src == jl_nothing) {
-            src = jl_type_infer(mi, world, 0);
-            if (src)
-                jlrettype = src->rettype;
-            else if (jl_is_method(mi->def.method)) {
-                src = mi->def.method->generator ? jl_code_for_staged(mi, world) : (jl_code_info_t*)mi->def.method->source;
-                if (src && (jl_value_t*)src != jl_nothing && !jl_is_code_info(src) && jl_is_method(mi->def.method))
-                    src = jl_uncompress_ir(mi->def.method, NULL, (jl_value_t*)src);
-            }
-            // TODO: use mi->uninferred
         }
     }
+    if (src) {
+        if ((jl_value_t*)src != jl_nothing && !jl_is_code_info(src) && jl_is_method(mi->def.method))
+            src = jl_uncompress_ir(mi->def.method, (jl_value_t*)src);
```

Comment:
Why is it fine to drop the handling of `@generated` functions here?

Reply:
I don't really know what that code was for. Inference will run that for you. If I missed something, it needs a testcase.

Reply:
It is for consistency with codegen, that if the generator succeeds, but `jl_type_infer` gets confused and errors (or I suppose if `jl_type_infer` decides to skip inference on it), then this result is still consistent with what codegen would do. But since this is mostly just a InteractiveUtils helpers, perhaps ideally we would move a lot this function into Julia instead, so that it was more of a reflection tool (like code_typed with no fallbacks) rather than a driver for codegen (e.g. acting like the JIT with fallbacks to preserve the dynamic semantics)